### PR TITLE
[KKS][ExtSave] Avoid error when saving more than 2 GB in the main game

### PR DESCRIPTION
**Pull request by takahiro0327**
Please review and merge.

https://github.com/IllusionMods/IllusionFixes/pull/63

### CODE REVIEW

File: src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs [hunk 1]
```diff
@@ -56,44 +56,67 @@ private static void SaveDataLoadHook(BinaryReader br, ref WorldData saveData)
...
+                        }
                     }
-                }
-                catch (Exception ex)
-                {
-                    UnityEngine.Debug.LogException(ex);
-                }
+                    catch (Exception ex)
+                    {
+                        UnityEngine.Debug.LogException(ex);
```
Comment (ManlyMarco) [inline src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs:106]:
```suggestion
                        Logger.Log(BepInEx.Logging.LogLevel.Message | BepInEx.Logging.LogLevel.Error, "Failed to save the game, the save file may be corrupted! Try saving again in an empty slot.\nError: " + ex.Message);
                        UnityEngine.Debug.LogException(ex);
```

File: src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs [hunk 2]
```diff
@@ -56,44 +56,67 @@ private static void SaveDataLoadHook(BinaryReader br, ref WorldData saveData)
...
+                return false;
+            }
+
+
+            [HarmonyPostfix, HarmonyPatch(typeof(WorldData), nameof(WorldData.GetBytes), typeof(WorldData))]
+            private static void SaveDataSaveHook(WorldData saveData, ref byte[] __result)
+            {
+                // This function should not be called.
+                // Originally called from SaveData.WorldData.Save(), but the original Save() is not called by the patch.
+                throw new System.NotSupportedException("Ported functionality to WorldData.Save patch.");
```
Comment (ManlyMarco) [inline src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs:119]:
```suggestion
                throw new System.NotSupportedException("Do not use this method, use WorldData.Save instead. More info: https://github.com/IllusionMods/BepisPlugins/pull/197");
```

File: src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs [hunk 3]
```diff
@@ -56,44 +56,67 @@ private static void SaveDataLoadHook(BinaryReader br, ref WorldData saveData)
...
-                            bw.Write(data);
-                        }
+                            // Append the ext data
+                            Dictionary<string, PluginData> extendedData = GetAllExtendedData(saveData);
 
-                        // Replace the result, not the most efficient way but save files shouldn't be big enough to matter since it's all done in memory
-                        __result = ms.ToArray();
+                            if (extendedData != null)
+                            {
+                                // Then write our data
```
Comment (ManlyMarco) [inline src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs:95]:
```suggestion
```

File: src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs [hunk 4]
```diff
@@ -56,44 +56,67 @@ private static void SaveDataLoadHook(BinaryReader br, ref WorldData saveData)
 
             #region Saving
 
-            // Nope, not patching the save lambda, not doing it, noooope
-            [HarmonyPostfix, HarmonyPatch(typeof(WorldData), nameof(WorldData.GetBytes), typeof(WorldData))]
-            private static void SaveDataSaveHook(WorldData saveData, ref byte[] __result)
+            [HarmonyPrefix, HarmonyPatch(typeof(SaveData.WorldData), nameof(SaveData.WorldData.Save), typeof(string), typeof(string))]
+            [HarmonyPriority(Priority.Last)]
+            static bool WorldDataSavePrefix(SaveData.WorldData __instance, string path, string fileName)
```
Comment (ManlyMarco) [inline src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs:61]:
```suggestion
            private static bool WorldDataSavePrefix(SaveData.WorldData __instance, string path, string fileName)
```

File: src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs [hunk 5]
```diff
@@ -56,44 +56,67 @@ private static void SaveDataLoadHook(BinaryReader br, ref WorldData saveData)
 
             #region Saving
 
-            // Nope, not patching the save lambda, not doing it, noooope
-            [HarmonyPostfix, HarmonyPatch(typeof(WorldData), nameof(WorldData.GetBytes), typeof(WorldData))]
-            private static void SaveDataSaveHook(WorldData saveData, ref byte[] __result)
+            [HarmonyPrefix, HarmonyPatch(typeof(SaveData.WorldData), nameof(SaveData.WorldData.Save), typeof(string), typeof(string))]
+            [HarmonyPriority(Priority.Last)]
+            private static bool WorldDataSavePrefix(SaveData.WorldData __instance, string path, string fileName)
```
Comment (ManlyMarco) [inline src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs:61]:
```suggestion
            private static bool WorldDataSaveOverride(SaveData.WorldData __instance, string path, string fileName)
```

File: src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs [hunk 6]
```diff
@@ -56,44 +56,67 @@ private static void SaveDataLoadHook(BinaryReader br, ref WorldData saveData)
...
+                        UnityEngine.Debug.LogException(ex);
+                    }
+                }));
+
+                return false;
+            }
+
+
+            [HarmonyPostfix, HarmonyPatch(typeof(WorldData), nameof(WorldData.GetBytes), typeof(WorldData))]
+            private static void SaveDataSaveHook(WorldData saveData, ref byte[] __result)
```
Comment (ManlyMarco) [inline src/KKS_ExtensibleSaveFormat/KKS.ExtendedSave.SaveData.Hooks.cs:115]:
```suggestion
            private static void WorldDataGetBytesDisableHook(WorldData saveData, ref byte[] __result)
```

**Comment by takahiro0327** [comment]
Thanks for the code fix and merge!

**Pull request merged**